subsys/llext/llext.c

```diff
@@ -709,48 +909,67 @@ int llext_load(struct llext_loader *ldr, const char *name, struct llext **ext)
 		}
 
 		ldr->hdr = ehdr;
-		ret = do_llext_load(ldr, *ext);
-		break;
-	default:
-		LOG_ERR("Unsupported elf file type %x", ehdr.e_type);
-		*ext = NULL;
-		return -EINVAL;
-	}
+		ret = do_llext_load(ldr, *ext, do_local);
+		if (ret < 0) {
+			k_heap_free(&llext_heap, *ext);
+			*ext = NULL;
+			return ret;
+		}
 
-	if (ret == 0) {
 		strncpy((*ext)->name, name, sizeof((*ext)->name));
 		(*ext)->name[sizeof((*ext)->name) - 1] = '\0';
+		(*ext)->use_count++;
+
 		sys_slist_append(&_llext_list, &(*ext)->_llext_list);
 		LOG_INF("Loaded extension %s", (*ext)->name);
+
+		break;
+	default:
+		LOG_ERR("Unsupported elf file type %x", ehdr.e_type);
+		return -EINVAL;
 	}
 
-	return ret;
+	return 0;
 }
 
-void llext_unload(struct llext *ext)
+void llext_unload(struct llext **ext, void (*free)(struct llext *, const void *), const void *ctx)
 {
-	__ASSERT(ext, "Expected non-null extension");
+	__ASSERT(*ext, "Expected non-null extension");
+	struct llext *tmp = *ext;
 
-	sys_slist_find_and_remove(&_llext_list, &ext->_llext_list);
+	k_mutex_lock(&llext_lock, K_FOREVER);
+	if (tmp->use_count-- != 1) {
+		k_mutex_unlock(&llext_lock);
+		return;
+	}
+
+	if (free) {
+		free(tmp, ctx);
```

Comment:
Could all the params given to unload for freeing be given to the loader and saved as part of the module instead? It's unclear otherwise how the free fn and ctx are to be used here to me.

Reply:
Agree with the above - even more, the use case for this feels unclear to me:
* Before calling `llext_unload` on an used llext, I expect the caller to have nothing more to do with the llext, so for a single user it feels redundant;
* Since now we have ref counting, if some llext is shared by 2 users, and each needs its own callback, this should be saved as a list so every user's callbacks get called. 

Still, I don't see a clear use for this - I keep thinking about calling special "init" and "exit" functions _in the llext_ (Linux module style)...

Reply:
@teburd I had that in my first version, but then I realised that those parameters are only used once here for freeing, and we do call `llext_unload()` explicitly anyway, so there's no point in storing those values. We can just pass them directly here.

Reply:
@pillo79 let me explain a bit. With SOF we have an auxiliary extension management layer. Its tasks include (1) loading extensions from an external source into slow but abundant memory, (2) copying them for use into fast but scarce memory, (3) freeing back that memory when no longer in use.
The flow is then as follows:
1. load from source during startup
2. when needed:
2.1. call `llext_load()` to parse and link ELF
2.2. check return code to see if it's the first use
2.3. if so, copy to fast RAM
3. when unneeded:
3.1. call `llext_unload()`
3.2. if that was the last use, our callback is called, which frees the fast memory

I think in step 3.2 above we could do the same as in 2.2 - check return code and free memory if last. Would that be a better solution? This also explains, that we don't have different callbacks for multiple users, but only one of them. It's also correct, that currently we don't use the `llext` context in our callback, but I thought that in the future that could change so I implemented it this way. But if we switch to return value checking, then, of course, that context will be gone anyway after `llext_unload()` returns.

Reply:
@pillo79 ok, I'll try to switch over to a return code then, stay tuned :-)